File: backend/classes/Configuration.php
<?php  

class Configuration{
	// TODO: remove this in case we switch to JSON or a better
	// configuration format
	public $config_file_variables;

	// database config
	public $db_server;
	public $db_name;
	public $db_user;
	public $db_password;

	// default user status / the ones that are shipped with the installation by default
	// Note: they have to match the IDs in the default
	//       schema definition
	public $default_guest_user_status_id = 1;
	public $default_member_user_status_id = 2;
	public $default_admin_user_status_id = 3;

	// user roles
	public $guest_user_role_id;
	public $member_user_role_id;
	public $admin_user_role_id;

	// properties
	public $browser_session_timeout_max;
	public $location_time_zone;
	public $location_longitude;
	public $location_latitude;
	public $business_day_start;
	public $business_day_end;
	public $session_cancel_graceperiod;
	public $recaptcha_publickey;
	public $recaptcha_privatekey;
	public $smtp_sender;
	public $smtp_server;
	public $smtp_username;
	public $smtp_password;

	// customization
	public $currency;
	public $location_map;
	public $location_address;
	public $payment_account_owner;
	public $payment_account_iban;
	public $payment_account_bic;
	public $payment_account_comment;
	public $logo_file;
	public $engine_hour_format;
	public $fuel_payment_type;


	// TODO: remove dependencies on these
	public $login_page;
	public $version;
	public $required_db_version;

    // Automatically load classes from the classes folder
	function __autoload($class_name){
		if(file_exists($class_name.'.php')){
			include $class_name.'.php';
		}
	}
	
	public function __construct(){
		// read the config file
		if(file_exists(dirname(__DIR__).'/config/config.php')){
			include dirname(__DIR__).'/config/config.php';
			$this->config_file_variables = $config;
		}else{
			//throw new Exception("config file does not exist");
			error_log("No config file found, return empty config");
			return NULL;
		}

		// get database configuration parameters
		// from the configuration file
		$this->get_configuration_file($this->config_file_variables);

		// get all the user roles from the database that
		// are required for the code to run
		$this->get_user_roles($this->config_file_variables);

		// get all the configuration settings from the
		// database
		$this->get_configuration_properties($this->config_file_variables);

		// set hard defaults
		$this->login_page          = '/index.html';
		$this->version             = "2.3.3";
		$this->required_db_version = "1.16";
	}

	public function is_db_configured(){
		// check if configuration is set
        if(!isset($this->db_server) or $this->db_server == ''){
            return FALSE;
        }
        // check if db_name is set
        if(!isset($this->db_name) or $this->db_name == ''){
            return FALSE;
        }
        // check if db_user is set
        if(!isset($this->db_user) or $this->db_user == ''){
            return FALSE;
        }

        return TRUE;
	}

	// Gets the content from the configuration file into
	// the public variables of this class
	private function get_configuration_file($config){
		$this->db_server   = $config['db_server'];
		$this->db_name     = $config['db_name'];
		$this->db_user     = $config['db_user'];
		$this->db_password = $config['db_password'];
	}
	
	// Returns a hash with the user's profile
	private function get_user_roles($config){
		// connect to the database
		$db = new DBAccess($this);
		if(!$db->connect()){
			error_log('classes/Configuration: Cannot connect to the database');
			return null;
		}
		
		// get different user_status from database (the default ones)
		$query = 'SELECT id, name FROM user_role;';
		$db->prepare($query);
		$db->execute();
		$res = $db->fetch_stmt_hash();
		$db->disconnect();
		if(! isset($res) or $res === FALSE){
			error_log('classes/Configuration: Cannot retrieve user roles');
			return $config;
		}

		// go through result and handle all supported user roles
		foreach($res as $user_role){
			switch($user_role['name']){
				case 'guest':
					$this->guest_user_role_id = $user_role['id'];
					break;
				case 'member':
					$this->member_user_role_id = $user_role['id'];
					break;
				case 'admin':
					$this->admin_user_role_id = $user_role['id'];
					break;
				default:
				    error_log("Unsupported user role: " . $user_role['name'] . " make sure this one is handled properly in the code");
					break;
			}
		}
		
		return;
	}

	private function get_configuration_properties($config){
		// connect to the database
		$db = new DBAccess($this);
		if(!$db->connect()){
			error_log('classes/Configuration: Cannot connect to the database');
			return null;
		}
		
		// get different user_status from database (the default ones)
		$query = 'SELECT property as name, value FROM configuration;';
		$db->prepare($query);
		$db->execute();
		$res = $db->fetch_stmt_hash();
		$db->disconnect();
		if(! isset($res) or $res === FALSE){
			error_log('classes/Configuration: Cannot retrieve user status');
			return $config;
		}

		// go through result
		foreach($res as $property){
			switch($property['name']){
				case 'browser.session.timeout.max': // TODO
					$this->browser_session_timeout_max = (int) $property['value'];
					break;
				case 'location.timezone':
					$this->location_time_zone = $property['value'];
					break;
				case 'location.longitude':
					$this->location_longitude = $property['value'];
					break;
				case 'location.latitude':
					$this->location_latitude = $property['value'];
					break;
				case 'business.day.start': // TODO
					$this->business_day_start = $property['value'];
					break;
				case 'business.day.end': // TODO
					$this->business_day_end = $property['value'];
					break;
				case 'session.cancel.graceperiod': // TODO
					$this->session_cancel_graceperiod = $property['value'];
					break;
				case 'recaptcha.publickey':
					$this->recaptcha_publickey = $property['value'];
					break;
				case 'recaptcha.privatekey':
					$this->recaptcha_privatekey = $property['value'];
					break;
				case 'currency':
					$this->currency = $property['value'];
					break;
				case 'location.map':
					$this->location_map = $property['value'];
					break;
				case 'location.address':
					$this->location_address = $property['value'];
					break;
				case 'payment.account.owner':
					$this->payment_account_owner = $property['value'];
					break;
				case 'payment.account.iban':
					$this->payment_account_iban = $property['value'];
					break;
				case 'payment.account.bic':
					$this->payment_account_bic = $property['value'];
					break;
				case 'payment.account.comment':
					$this->payment_account_comment = $property['value'];
					break;
				case 'smtp.sender':
					$this->smtp_sender = $property['value'];
					break;
				case 'smtp.server': 
					$this->smtp_server = $property['value'];
					break;
				case 'smtp.username':
					$this->smtp_username = $property['value'];
					break;
				case 'smtp.password':
					$this->smtp_password = $property['value'];
					break;
				case 'logo.file':
					$this->logo_file = $property['value'];
					break;
				case 'engine.hour.format':
					$this->engine_hour_format = $property['value'];
					break;
				case 'fuel.payment.type':
					$this->fuel_payment_type = $property['value'];
			}
		}

		// set defaults for values that are not specified yet
		if(is_null($this->browser_session_timeout_max)){
			$this->browser_session_timeout_max = 604800;
		}
		if(is_null($this->location_time_zone)){
			$this->location_time_zone = "Europe/Berlin";
		}
		if(is_null($this->location_longitude)){
			$this->location_longitude = 8.542939;
		}
		if(is_null($this->location_latitude)){
			$this->location_latitude = 47.367658;
		}
		if(is_null($this->business_day_start)){
			$this->business_day_start = "08:00:00";
		}
		if(is_null($this->business_day_end)){
			$this->business_day_end = "21:00:00";
		}
		if(is_null($this->session_cancel_graceperiod)){
			$this->session_cancel_graceperiod = 86400;
		}
		if(is_null($this->currency)){
			$this->currency = "CHF";
		}
	}

	public function set_configuration_property($key, $value){
		$db = new DBAccess($this);
        if(!$db->connect()){
			error_log("Cannot connect to database");
            return FALSE;
        }
		$db->prepare('REPLACE INTO configuration (property, value) VALUES (?, ?);');
		
		$db_key = '';
		switch ($key) {
			case 'engine_hour_format':
				$db_key = "engine.hour.format";
				break;
			case 'fuel_payment_type':
				$db_key = "fuel.payment.type";
				break;
			case 'logo_file':
				$db_key = "logo.file";
				break;
			case 'location_time_zone':
				$db_key = "location.timezone";
				break;
			case 'location_latitude':
				$db_key = "location.latitude";
				break;
			case 'location_longitude':
				$db_key = "location.longitude";
				break;
			case 'location_map':
				$db_key = "location.map";
				break;
			case 'location_address':
				$db_key = "location.address";
				break;
			case 'currency':
				$db_key = "currency";
				break;
			case 'payment_account_owner':
				$db_key = "payment.account.owner";
				break;
			case 'payment_account_iban':
				$db_key = "payment.account.iban";
				break;
			case 'payment_account_bic':
				$db_key = "payment.account.bic";
				break;
			case 'payment_account_comment':
				$db_key = "payment.account.comment";
				break;
			case 'smtp_sender':
				$db_key = "smtp.sender";
				break;
			case 'smtp_server':
				$db_key = "smtp.server";
				break;
			case 'smtp_username':
				$db_key = "smtp.username";
				break;
			case 'smtp_password':
				$db_key = "smtp.password";
				break;
			case 'recaptcha_privatekey':
				$db_key = "recaptcha.privatekey";
				break;
			case 'recaptcha_publickey':
				$db_key = "recaptcha.publickey";
				break;
			default:
				return FALSE;
		}

		$db->bind_param('ss',
			$db_key,
			$value
		);
		if(! $db->execute()){
			return FALSE;
		}
		return TRUE;
	}
}
?>

File: backend/api/v2/handlers/configuration.go
package handlers

import (
	"crypto/sha256"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"server/database"
	"strconv"
	"time"

	"golang.org/x/exp/slog"
)

// type GetDBConfigResponse struct {
// 	IsConfigured bool   `json:"is_configured"`
// 	DBServer     string `json:"db_server"`
// 	DBName       string `json:"db_name"`
// 	DBUser       string `json:"db_user"`
// 	DBPassword   string `json:"db_password,omitempty"`
// }

// PublicConfigurationMessage represents the public configuration of the server.
// It is available to regular users that are logged in.
type PublicConfigurationMessage struct {
	Currency               string  `json:"currency" validate:"required,excludesall={} []!()<>"`
	EngineHourFormat       string  `json:"engine_hour_format" validate:"required,oneof=hh.h hh:mm"`    // check if really needed
	FuelPaymentType        string  `json:"fuel_payment_type" validate:"required,oneof=billed instant"` // check if really needed
	LocationAddress        string  `json:"location_address" validate:"excludesall={}[]!><"`
	LocationLatitude       float32 `json:"location_latitude" validate:"required,latitude"`
	LocationLongitude      float32 `json:"location_longitude" validate:"required,longitude"`
	LocationMap            string  `json:"location_map" validate:"omitempty,googlemapsurl"`
	LocationTimeZone       string  `json:"location_time_zone" validate:"required"`
	LogoFilePath           string  `json:"logo_file" validate:"omitempty,filepath"`                                                      // TODO: validator for uploaded file | check if really needed
	MyNautiqueEnabled      bool    `json:"mynautique_enabled" validate:"omitempty,boolean"`                                              // check if really needed
	MyNautiqueFuelCapacity int     `json:"mynautique_fuel_capacity" validate:"required_if=MyNautiqueEnabled true,omitempty,number,gt=1"` // check if really needed
	MyNautiqueBoatID       int     `json:"mynautique_boat_id" validate:"required_if=MyNautiqueEnabled true,omitempty,number,gt=1"`       // required to query mynautique telemetry
	PaymentAccountBIC      string  `json:"payment_account_bic" validate:"omitempty,printascii"`
	PaymentAccountComment  string  `json:"payment_account_comment"`
	PaymentAccountIBAN     string  `json:"payment_account_iban" validate:"omitempty,printascii"`
	PaymentAccountOwner    string  `json:"payment_account_owner"`
	RecaptchaPublicKey     string  `json:"recaptcha_publickey" validate:"omitempty,recaptchakey,required_with=RecaptchaPrivateKey"` // check if really needed
}

type ConfigurationMessage struct {
	Currency               string  `json:"currency" validate:"required,excludesall={} []!()<>"`
	EngineHourFormat       string  `json:"engine_hour_format" validate:"required,oneof=hh.h hh:mm"`
	FuelPaymentType        string  `json:"fuel_payment_type" validate:"required,oneof=billed instant"`
	LocationAddress        string  `json:"location_address" validate:"excludesall={}[]!><"`
	LocationLatitude       float32 `json:"location_latitude" validate:"required,latitude"`
	LocationLongitude      float32 `json:"location_longitude" validate:"required,longitude"`
	LocationMap            string  `json:"location_map" validate:"omitempty,googlemapsurl"`
	LocationTimeZone       string  `json:"location_time_zone" validate:"required"`
	LogoFilePath           string  `json:"logo_file" validate:"omitempty,filepath"` // TODO: validator for uploaded file
	MyNautiqueBoatID       int     `json:"mynautique_boat_id" validate:"required_if=MyNautiqueEnabled true,omitempty,number,gt=10"`
	MyNautiqueEnabled      bool    `json:"mynautique_enabled" validate:"omitempty,boolean"`
	MyNautiqueFuelCapacity int     `json:"mynautique_fuel_capacity" validate:"required_if=MyNautiqueEnabled true,omitempty,number,gt=10"`
	MyNautiquePassword     string  `json:"mynautique_password" validate:"required_if=MyNautiqueEnabled true,omitempty,gt=1"`
	MyNautiqueUser         string  `json:"mynautique_user" validate:"required_if=MyNautiqueEnabled true,omitempty,email"`
	PaymentAccountBIC      string  `json:"payment_account_bic" validate:"omitempty,printascii"`
	PaymentAccountComment  string  `json:"payment_account_comment"`
	PaymentAccountIBAN     string  `json:"payment_account_iban" validate:"omitempty,printascii"`
	PaymentAccountOwner    string  `json:"payment_account_owner"`
	RecaptchaPrivateKey    string  `json:"recaptcha_privatekey" validate:"omitempty,recaptchakey,required_with=RecaptchaPublicKey"`
	RecaptchaPublicKey     string  `json:"recaptcha_publickey" validate:"omitempty,recaptchakey,required_with=RecaptchaPrivateKey"`
	SMTPPassword           string  `json:"smtp_password" validate:"omitempty"`
	SMTPSender             string  `json:"smtp_sender" validate:"omitempty,required_with=SMTPSender,email"`
	SMTPServer             string  `json:"smtp_server" validate:"required_with=SMTPSender"`
	SMTPUsername           string  `json:"smtp_username" validate:"required_with=SMTPSender"`
}

type ConfigSource int

const (
	SourceCLI ConfigSource = iota
	SourceDatabase
)

type StringConfigValue struct {
	Value  string
	Source ConfigSource
}

var ConfigurationMessageValidationErrors = map[string]string{
	"Currency":               "Use the 3 letter currency representation. E.g., USD, EUR, CHF.",
	"EngineHourFormat":       "Engine hour format can only be hh.m or hh:mm.",
	"FuelPaymentType":        "Fuel payment type is either 'instant' or 'billed'",
	"LocationAddress":        "Location address cannot contain invalid characters such as []{}<> or similar.",
	"LocationLatitude":       "Latitude is not a valid value.",
	"LocationLongitude":      "Longitude is not a valid value.",
	"LocationMap":            "Map URL needs to be a google embeded maps URL of the form https://www.google.com/maps/embeded?pb=...",
	"LocationTimeZone":       "The timezone needs to be a valid representation such as Europe/Zurich.",
	"LogoFilePath":           "Needs to be the path to the logo file on the server.",
	"MyNautiqueAPIKey":       "The API key for the my Nautique App needs to be provided.",
	"MyNautiqueBoatID":       "The boat ID from the my Nautique App should be a number.",
	"MyNautiqueEnabled":      "MyNautique enabled needs to be true or false",
	"MyNautiqueFuelCapacity": "The MyNautique fuel capacity needs to be a number.",
	"MyNautiqueUser":         "The user for MyNautique needs to be an Email address.",
	"MyNautiquePassword":     "A password for the MyNautique user needs to be set.",
	"PaymentAccountBIC":      "BIC needs to be set to a valid value.",
	"PaymentAccountComment":  "Comment cannot be parsed.",
	"PaymentAccountIBAN":     "The IBAN number needs to be valid.",
	"PaymentAccountOwner":    "The account owner cannot be parsed.",
	"RecaptchaPrivateKey":    "Recaptcha private key is not valid.",
	"RecaptchaPublicKey":     "Recaptcha public key is not valid.",
	"SMTPPassword":           "SMTP Password needs to be set.",
	"SMTPSender":             "SMTP sender is not a valid email address.",
	"SMTPServer":             "SMTP server is not a valid  address.",
	"SMTPUsername":           "No or invalid SMTP username provided",
}

type SetupDBConfigRequest struct {
	DBServer   string `json:"db_server" validate:"required,hostname_port"`
	DBName     string `json:"db_name" validate:"required,alphanum"`
	DBUser     string `json:"db_user" validate:"required,alphanum"`
	DBPassword string `json:"db_password" validate:"required"`
}

type SetupMyNautiqueCredentialsRequest struct {
	Enabled  bool   `json:"mynautique_enabled" validate:"boolean"`
	User     string `json:"mynautique_user" validate:"omitempty,email"`
	Password string `json:"mynautique_password"`
}

type UploadLogoFileResponse struct {
	URI      string `json:"uri"`
	FileName string `json:"filename"`
}

type GetLogoPathResponse struct {
	URI string `json:"uri"`
}

type GetRecaptchaKeyResponse struct {
	Key string `json:"key"`
}

func (h *Handler) SetupDBConfig(w http.ResponseWriter, r *http.Request) {
	configFile, _ := h.config.GetString("config")
	if configFile == "" {
		slog.Warn("No config file path provided to store configuration. Database setup not possible.")
		WriteFailureResponse("No config file path provided to store configuration. Database setup not possible, please provide a configuration file (--config flag).", w)
		return
	}
	// We only want to allow setup for the DB, if DB settings are not
	// already present in the configuration.
	if h.config.IsDBConfigured() {
		slog.Warn("SetupDB called but database settings are already configured.")
		WriteFailureResponse("Invalid request. Database config was created already and cannot be overwritten that way.", w)
		return
	}

	var req SetupDBConfigRequest
	err := ReadBodyAndValidate(r, &req)
	if err != nil {
		slog.Warn("Request payload is not valid", slog.String("error", err.Error()))
		WriteFailureResponse("Invalid request payload", w)
		return
	}

	host, port, err := net.SplitHostPort(req.DBServer)
	if err != nil {
		slog.Warn("Request payload is not valid, invalid DBServer", slog.String("error", err.Error()))
		WriteFailureResponse("Invalid hostname (should be host:port)", w)
		return
	}

	// Connect to database and initialize.
	err = h.Database.ConnectAndReplace(&database.Settings{
		User:     req.DBUser,
		Password: req.DBPassword,
		Protocol: "tcp",
		Host:     host,
		Port:     port,
		DBName:   req.DBName,
	})
	if err != nil {
		slog.Warn(fmt.Sprintf("New database parameters are not valid. Error returned from Connet(): %s", err))
		WriteFailureResponse("Cannot connect to database. Please make sure that the credentials are correct and the database is accepting connections.", w)
		return
	}

	// If database access was successful, we store the configuration.
	h.config.SetConfigFileValue("database.user", req.DBUser)
	h.config.SetConfigFileValue("database.password", req.DBPassword)
	h.config.SetConfigFileValue("database.protocol", "tcp")
	h.config.SetConfigFileValue("database.host", host)
	h.config.SetConfigFileValue("database.port", port)
	h.config.SetConfigFileValue("database.dbname", req.DBName)
	err = h.config.WriteConfigFile()
	if err != nil {
		slog.Warn("Cannot store new configuration", slog.String("error", err.Error()))
		WriteFailureResponse("Database was setup. No config file path provided to store database configuration, please start application by providing a config file.", w)
		return
	}
	slog.Info("New database configuration has been written to", slog.String("config", configFile))
	// TODO: The database manager should just be calles with a ConnectAndReplace like above
	// and we should be good and not need this part here.
	h.config.SetDB(h.Database)
	WriteSuccessResponse("config written", nil, w)
}

func (h *Handler) GetPublicConfiguration(w http.ResponseWriter, r *http.Request) {
	currency, _ := h.config.GetString("currency")
	engineHourFormat, _ := h.config.GetString("engine.hour.format")
	fuelPaymentType, _ := h.config.GetString("fuel.payment.type")
	locationAddress, _ := h.config.GetString("location.address")
	lat, _ := h.config.GetString("location.latitude")
	lon, _ := h.config.GetString("location.longitude")
	lat32, _ := strconv.ParseFloat(lat, 32)
	lon32, _ := strconv.ParseFloat(lon, 32)
	locationMap, _ := h.config.GetString("location.map")
	locationTimeZone, _ := h.config.GetString("location.timezone")
	logoFilePath, _ := h.config.GetString("logo.file")
	myNautiqueEnabled := h.config.GetBool("mynautique.enabled")
	myNautiqueBoatID := h.config.GetInt64("mynautique.boat.id")
	myNautiqueFuelCapacity := h.config.GetInt64("mynautique.fuel.capacity")
	paymentAccountBIC, _ := h.config.GetString("payment.account.bic")
	paymentAccountComment, _ := h.config.GetString("payment.account.comment")
	paymentAccountIBAN, _ := h.config.GetString("payment.account.iban")
	paymentAccountOwner, _ := h.config.GetString("payment.account.owner")
	recaptchePublicKey, _ := h.config.GetString("recaptcha.publickey")
	resp := &PublicConfigurationMessage{
		Currency:               currency,
		EngineHourFormat:       engineHourFormat,
		FuelPaymentType:        fuelPaymentType,
		LocationAddress:        locationAddress,
		LocationLatitude:       float32(lat32),
		LocationLongitude:      float32(lon32),
		LocationMap:            locationMap,
		LocationTimeZone:       locationTimeZone,
		LogoFilePath:           logoFilePath,
		MyNautiqueEnabled:      myNautiqueEnabled,
		MyNautiqueBoatID:       int(myNautiqueBoatID),
		MyNautiqueFuelCapacity: int(myNautiqueFuelCapacity),
		PaymentAccountBIC:      paymentAccountBIC,
		PaymentAccountComment:  paymentAccountComment,
		PaymentAccountIBAN:     paymentAccountIBAN,
		PaymentAccountOwner:    paymentAccountOwner,
		RecaptchaPublicKey:     recaptchePublicKey,
	}
	WriteSuccessResponse("configuration", resp, w)
}

// TODO create public and non public version of this
// E.g. Public version should not contain private info like mynautique, db, ... things.
func (h *Handler) GetConfiguration(w http.ResponseWriter, r *http.Request) {
	currency, _ := h.config.GetString("currency")
	engineHourFormat, _ := h.config.GetString("engine.hour.format")
	fuelPaymentType, _ := h.config.GetString("fuel.payment.type")
	locationAddress, _ := h.config.GetString("location.address")
	lat, _ := h.config.GetString("location.latitude")
	lon, _ := h.config.GetString("location.longitude")
	lat32, _ := strconv.ParseFloat(lat, 32)
	lon32, _ := strconv.ParseFloat(lon, 32)
	locationMap, _ := h.config.GetString("location.map")
	locationTimeZone, _ := h.config.GetString("location.timezone")
	logoFilePath, _ := h.config.GetString("logo.file")
	myNautiqueEnabled := h.config.GetBool("mynautique.enabled")
	myNautiqueBoatID := h.config.GetInt64("mynautique.boat.id")
	myNautiqueFuelCapacity := h.config.GetInt64("mynautique.fuel.capacity")
	myNautiqueUser, _ := h.config.GetString("mynautique.user")
	paymentAccountBIC, _ := h.config.GetString("payment.account.bic")
	paymentAccountComment, _ := h.config.GetString("payment.account.comment")
	paymentAccountIBAN, _ := h.config.GetString("payment.account.iban")
	paymentAccountOwner, _ := h.config.GetString("payment.account.owner")
	recaptchePublicKey, _ := h.config.GetString("recaptcha.publickey")
	recaptchePrivateKey, _ := h.config.GetString("recaptcha.privatekey")
	smtpSender, _ := h.config.GetString("smtp.sender")
	smtpServer, _ := h.config.GetString("smtp.server")
	smtpUsername, _ := h.config.GetString("smtp.username")

	resp := &ConfigurationMessage{
		Currency:               currency,
		EngineHourFormat:       engineHourFormat,
		FuelPaymentType:        fuelPaymentType,
		LocationAddress:        locationAddress,
		LocationLatitude:       float32(lat32),
		LocationLongitude:      float32(lon32),
		LocationMap:            locationMap,
		LocationTimeZone:       locationTimeZone,
		LogoFilePath:           logoFilePath,
		MyNautiqueBoatID:       int(myNautiqueBoatID),
		MyNautiqueEnabled:      myNautiqueEnabled,
		MyNautiqueFuelCapacity: int(myNautiqueFuelCapacity),
		MyNautiquePassword:     "hidden",
		MyNautiqueUser:         myNautiqueUser,
		PaymentAccountBIC:      paymentAccountBIC,
		PaymentAccountComment:  paymentAccountComment,
		PaymentAccountIBAN:     paymentAccountIBAN,
		PaymentAccountOwner:    paymentAccountOwner,
		RecaptchaPrivateKey:    recaptchePrivateKey,
		RecaptchaPublicKey:     recaptchePublicKey,
		SMTPPassword:           "hidden",
		SMTPSender:             smtpSender,
		SMTPServer:             smtpServer,
		SMTPUsername:           smtpUsername,
	}
	// TODO return if we have a myNautique API key. This is to
	// decide whether to show the myNautique settings section in the
	// UI.
	WriteSuccessResponse("configuration", resp, w)
}

func (h *Handler) SetConfiguration(w http.ResponseWriter, r *http.Request, req ConfigurationMessage, hCtx *HandlerCtx) {
	// TODO check smtp.password
	// - needs to be set in case there is other smtp configuration
	// - can be empty in case there is a password in the db already
	// - empty will not change the password in the db

	props := []database.Configuration{
		{
			Property: "currency",
			Value:    req.Currency,
		},
		{
			Property: "engine.hour.format",
			Value:    req.EngineHourFormat,
		},
		{
			Property: "fuel.payment.type",
			Value:    req.FuelPaymentType,
		},
		{
			Property: "location.address",
			Value:    req.LocationAddress,
		},
		{
			Property: "location.latitude",
			Value:    strconv.FormatFloat(float64(req.LocationLatitude), 'f', 6, 32),
		},
		{
			Property: "location.longitude",
			Value:    strconv.FormatFloat(float64(req.LocationLongitude), 'f', 6, 32),
		},
		{
			Property: "location.map",
			Value:    req.LocationMap,
		},
		{
			Property: "location.timezone",
			Value:    req.LocationTimeZone,
		},
		{
			Property: "logo.file",
			Value:    req.LogoFilePath,
		},
		{
			Property: "mynautique.boat.id",
			Value:    strconv.FormatInt(int64(req.MyNautiqueBoatID), 10),
		},
		{
			Property: "mynautique.enabled",
			Value:    strconv.FormatBool(req.MyNautiqueEnabled),
		},
		{
			Property: "mynautique.fuel.capacity",
			Value:    strconv.FormatInt(int64(req.MyNautiqueFuelCapacity), 10),
		},
		{
			Property: "mynautique.password",
			Value:    req.MyNautiquePassword,
		},
		{
			Property: "mynautique.user",
			Value:    req.MyNautiqueUser,
		},
		{
			Property: "payment.account.bic",
			Value:    req.PaymentAccountBIC,
		},
		{
			Property: "payment.account.comment",
			Value:    req.PaymentAccountComment,
		},
		{
			Property: "payment.account.iban",
			Value:    req.PaymentAccountIBAN,
		},
		{
			Property: "payment.account.owner",
			Value:    req.PaymentAccountOwner,
		},
		{
			Property: "recaptcha.privatekey",
			Value:    req.RecaptchaPrivateKey,
		},
		{
			Property: "recaptcha.publickey",
			Value:    req.RecaptchaPublicKey,
		},
		{
			Property: "smtp.password",
			Value:    req.SMTPPassword,
		},
		{
			Property: "smtp.sender",
			Value:    req.SMTPSender,
		},
		{
			Property: "smtp.server",
			Value:    req.SMTPServer,
		},
		{
			Property: "smtp.username",
			Value:    req.SMTPUsername,
		},
	}
	dbh := hCtx.Database
	err := dbh.UpdateOrInsertPropertyValues(props)
	if err != nil {
		slog.Warn("Cannot update configuration", slog.String("error", err.Error()))
		WriteFailureResponse(err.Error(), w)
		return
	}

	WriteSuccessResponse("config updated", nil, w)
}

func (h *Handler) SetupMyNautiqueCredentials(w http.ResponseWriter, r *http.Request, req SetupMyNautiqueCredentialsRequest, _ *HandlerCtx) {
	// in case the configuration is already present, we do not
	// allow to edit it
	if h.config.IsSet("mynautique.enabled") {
		slog.Warn("mynautique.enabled is already set, not allowing call to setup my nautique credentials.")
		return
	}

	err := h.config.SetPropertyValue("mynautique.enabled", fmt.Sprintf("%t", req.Enabled))
	if err != nil {
		slog.Warn("Request payload is not valid", slog.String("error", err.Error()))
		WriteFailureResponse("Invalid request payload", w)
		return
	}
	err = h.config.SetPropertyValue("mynautique.user", req.User)
	if err != nil {
		slog.Warn("Request payload is not valid", slog.String("error", err.Error()))
		WriteFailureResponse("Invalid request payload", w)
		return
	}
	err = h.config.SetPropertyValue("mynautique.password", req.Password)
	if err != nil {
		slog.Warn("Request payload is not valid", slog.String("error", err.Error()))
		WriteFailureResponse("Invalid request payload", w)
		return
	}

	WriteSuccessResponse("config written", nil, w)
}

func (h *Handler) UploadLogoFile(w http.ResponseWriter, r *http.Request) {
	file, fileHeader, err := r.FormFile("logo")
	if err != nil {
		slog.Warn("cannot access the form file", slog.String("position", "logo"))
		WriteFailureResponse("Cannot read the uploaded file.", w)
		return
	}
	defer file.Close()

	// check for the max logo size to be 512kB
	if fileHeader.Size > 512*1024 {
		slog.Warn("attempt to upload a file larger than 512kB", slog.Int64("size", fileHeader.Size))
		WriteFailureResponse("File is larger than 512kB.", w)
		return
	}

	contentTypes := fileHeader.Header.Values("Content-Type")
	if len(contentTypes) == 0 {
		slog.Warn("Cannot read Content-Type of file")
		WriteFailureResponse("The uploaded file needs to be an image (e.g., Content-Type image/jpeg)", w)
		return
	}
	contentType := contentTypes[0]
	if contentType != "image/png" &&
		contentType != "image/jpeg" &&
		contentType != "image/jpg" &&
		contentType != "image/gif" {
		slog.Warn("Wrong Content-Type of file", slog.String("content_type", contentType))
		WriteFailureResponse("The uploaded file needs to be an image (e.g., Content-Type image/jpeg)", w)
		return
	}

	// store uploaded file into local path. We use a content based filename
	// to store multiple files without collisions.
	nameTime := fmt.Sprintf("%s_%s", time.Now().Format(time.RFC3339), fileHeader.Filename)
	hash := sha256.Sum256([]byte(nameTime))
	fileHash := fmt.Sprintf("%x%s", hash[:16], filepath.Ext(fileHeader.Filename))
	storageDir, _ := h.config.GetString("http.uploadpath")
	localFileName := filepath.Join(storageDir, fileHash)
	if err := os.MkdirAll(filepath.Dir(localFileName), 0770); err != nil {
		slog.Warn("Cannot create directory for file", slog.String("file", localFileName), slog.String("error", err.Error()))
		WriteFailureResponse("File cannot get stored on server.", w)
		return
	}
	out, err := os.Create(localFileName)
	if err != nil {
		slog.Warn("Cannot create file on server for file", slog.String("file", localFileName), slog.String("error", err.Error()))
		WriteFailureResponse("File cannot get stored on server.", w)
		return
	}
	defer out.Close()
	_, err = io.Copy(out, file)
	if err != nil {
		slog.Warn("Cannot write to file on server for file", slog.String("file", localFileName), slog.String("error", err.Error()))
		WriteFailureResponse("File cannot get stored on server.", w)
		return
	}

	resp := &UploadLogoFileResponse{
		URI:      localFileName,
		FileName: fileHash,
	}
	WriteSuccessResponse("file uploaded", resp, w)
}

func (h *Handler) GetLogoPath(w http.ResponseWriter, r *http.Request) {
	conf, _ := h.config.GetString("logo.file")
	uploadDir, _ := h.config.GetString("http.uploadpath")
	slog.Warn("DEBUG: uploadDir", slog.String("dir", uploadDir))
	resp := &GetLogoPathResponse{}
	if conf != "" {
		resp.URI = filepath.Join(uploadDir, conf)
	}
	WriteSuccessResponse("logo path", resp, w)
}

// TODO implement file removal

func (h *Handler) GetRecaptchaKey(w http.ResponseWriter, r *http.Request) {
	key, _ := h.config.GetString("recaptcha.publickey")
	resp := &GetRecaptchaKeyResponse{}
	resp.Key = key
	WriteSuccessResponse("recaptcha key", resp, w)
}
